### ISSUE TITLE
vector<l1t::Muon> read too many bytes: 187 instead of 171

### ISSUE THREAD

**Issue by silviodonato** [cms-sw/cmssw#32133]
From https://hypernews.cern.ch/HyperNews/CMS/get/tier0-Ops/2187/1/1/1.html and
https://docs.google.com/document/d/1LWs-87dWMTlIj2VhmVZLcPhqZEGZK1Zkv1fRLeBbO1E/edit?pli=1#

In `CMSSW_11_1_5`, we get
```
----- Begin Fatal Exception 13-Nov-2020 14:52:39 CET-----------------------
An exception of category 'FatalRootError' occurred while
   [0] Calling InputSource::getNextItemType
   Additional Info:
      [a] Fatal Root Error: @SUB=TBufferFile::CheckByteCount
object of class vector<l1t::Muon> read too many bytes: 187 instead of 171

----- End Fatal Exception -------------------------------------------------
```

You can reproduce the error by 
```
cmsRun /afs/cern.ch/user/s/sdonato/AFSwork/public/L1issue/PSetPklAsPythonFile.py
```

**Comment by silviodonato** [comment]
assign l1

**Comment by silviodonato** [comment]
cc @boudoul @amassiro

**Comment by silviodonato** [comment]
I can confirm that `CMSSW_11_1_4` works without error. The error appears running with `CMSSW_11_1_4` + https://github.com/cms-sw/cmssw/pull/31679

**Comment by makortel** [comment]
The `ioread` in https://github.com/cms-sw/cmssw/pull/31679 11_1_X backport has a slight difference wrt. 11_2_X original PR https://github.com/cms-sw/cmssw/pull/31608 (red #31608, green #31679)
```diff
-  <ioread sourceClass="l1t::RegionalMuonCand" targetClass="l1t::RegionalMuonCand" version="[-10]"  source="int m_hwPt2" target="m_hwPtUnconstrained">
+  <ioread sourceClass="l1t::RegionalMuonCand" targetClass="l1t::RegionalMuonCand" version="[10]"  source="int m_hwPt2" target="m_hwPtUnconstrained">
```

**Comment by silviodonato** [comment]
> The `ioread` in #31679 11_1_X backport has a slight difference wrt. 11_2_X original PR #31608 (red #31608, green #31679)
> 
> ```diff
> -  <ioread sourceClass="l1t::RegionalMuonCand" targetClass="l1t::RegionalMuonCand" version="[-10]"  source="int m_hwPt2" target="m_hwPtUnconstrained">
> +  <ioread sourceClass="l1t::RegionalMuonCand" targetClass="l1t::RegionalMuonCand" version="[10]"  source="int m_hwPt2" target="m_hwPtUnconstrained">
> ```

ah! we need to backport #31848. Let's see if it actually solves the problem.

**Comment by boudoul** [comment]
Thanks! Adding @dinyar as watcher

**Comment by dinyar** [comment]
Hi all,

Thanks for looking into this! Regarding the above, I think it's a red herring unless I'm overlooking something. #31608 didn't contain @silviodonato's fix, but #31679 does. We in fact want `[10]` and not `[-10]`. Also the DF affected by this is RegionalMuonCand and not Muon. I'll try to reproduce locally.

Cheers,
Dinyar

**Comment by silviodonato** [comment]
ah, no, it was already backport with https://github.com/cms-l1t-offline/cmssw/commit/c3734a8d729369719a64077b5e6f4571ec6d77b3

so  #31679 is the backport of #31608 +  #31848

indeed we have 
```
<ioread sourceClass="l1t::RegionalMuonCand" targetClass="l1t::RegionalMuonCand" version="[10]"  source="int m_hwPt2" target="m_hwPtUnconstrained">
```

both in 
https://github.com/cms-sw/cmssw/blob/CMSSW_11_1_5/DataFormats/L1TMuon/src/classes_def.xml#L8
and in
https://github.com/cms-sw/cmssw/blob/master/DataFormats/L1TMuon/src/classes_def.xml#L8

**Comment by makortel** [comment]
Ah, right, sorry for the noise.

**Comment by silviodonato** [comment]
I've investigated a bit the problem. From the HLT menu used (`hltGetConfiguration run:337973 > hlt.py`) I found that `ALCAPHYSIM` and `NanoDST` are the only streams storing `hltGtStage2Digis`. `ALCAPHYSIM` triggered no data, so we get no error in that case (see `AlCa_EcalPhiSym_v4 ` in https://cmsoms.cern.ch/cms/triggers/hlt_report?cms_run=337973).
This is the reason why we get error only in `NanoDST`.

The failing process converts the `.dat` files into the RAW `.root` files. In case of `NanoDST` the RAW dataformat contains:
```
FEDRawDataCollection        "hltFEDSelector"     ""          "HLT"     
edm::TriggerResults         "TriggerResults"     ""          "HLT" 
BXVector<l1t::Muon>         "hltGtStage2Digis"   "Muon"      "HLT"     
BXVector<l1t::Tau>          "hltGtStage2Digis"   "Tau"       "HLT"     
[...]
```

Since we have changed the `l1t::Muon` dataformat, the expected length of the `l1t::Muon` size in the `.dat` file has changed causing the error. I expect that if we convert from `.root` to `.dat` and then again `.root` in `CMSSW_11_1_5`, everything should work. In other words, I do not expect we will get any crash during the MWGR. But I let @cms-sw/daq-l2 comment more on this.
